### PR TITLE
Fix NavigationTarget ranges

### PR DESCRIPTION
**Pull request by edwin0cheng**
Fix the issue described in https://github.com/rust-analyzer/rust-analyzer/pull/2544#issuecomment-565572553

This PR change the order for finding `full_range` of `focus_range` in following orders:
1. map both ranges to macro_call
2. map focus range to a token inside macro call, and full range to the whole of macro call
3. map both ranges to the whole of macro call

And fix the corresponding tests and make these tests easily to follow.

### CODE REVIEW

File: crates/ra_hir_expand/src/lib.rs [hunk 1]
```diff
@@ -214,7 +214,17 @@ pub struct ExpansionInfo {
     exp_map: Arc<mbe::TokenMap>,
 }
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ExpansionOrigin {
```
Comment (matklad) [inline crates/ra_hir_expand/src/lib.rs]: Let's just re-export `mbe::Origin`

File: crates/ra_ide/src/display/navigation_target.rs [hunk 2]
```diff
@@ -416,3 +420,14 @@ pub(crate) fn description_from_symbol(db: &RootDatabase, symbol: &FileSymbol) ->
         }
     }
 }
+
+fn original_range(db: &RootDatabase, node: InFile<&SyntaxNode>) -> FileRange {
+    if let Some(range) = original_range_by_kind(db, node, OriginalRangeKind::CallToken) {
```
Comment (matklad) [inline crates/ra_ide/src/display/navigation_target.rs]: I am not sure we need this more flexible `original_range_by_kind` function. I think the logic should be as follows:

* we try to map range back
* if we fail completely, we return the range of the macro_call
* if we land inside macro_argument, we return that range
* if we land inside macro_def, we return the range of the macro_call

Comment (edwin0cheng) [reply crates/ra_ide/src/display/navigation_target.rs]: My reason for adding this function is, do `original_range` only used for `NavigationTarget` ? (And that's my reason to add `Any` even it is not used for now)

I have a feeling we would used it in other place, which we need to mapping the range inside `MacroDef`.

Comment (edwin0cheng) [reply crates/ra_ide/src/display/navigation_target.rs]: Um.. Ah, I see what you mean right now, you mean we return the origin of what `original_source` find, and do the logic you mentioned above. It is functionally equal to my approach but much cleaner..

**Comment by edwin0cheng** [comment]
Fixed, but I rearrange the logic to:

1. if we land inside macro_argument, we return that range
2. we return the range of the macro_call

**Pull request merged**